test/e2e_node/cpu_manager_test.go

```diff
@@ -906,6 +796,149 @@ var _ = SIGDescribe("CPU Manager [Serial] [Feature:CPUManager]", func() {
+				_, allocatableCPU, _ = getLocalNodeCPUDetails(ctx, f)
+
+				// strict SMT alignment is trivially verified and granted on non-SMT systems
+				if smtLevel < 2 {
+					e2eskipper.Skipf("Skipping CPU Manager %s tests since SMT disabled", fullCPUsOnlyOpt)
+				}
+
+				// our tests want to allocate a full core, so we need at last 2*2=4 virtual cpus
+				if allocatableCPU < int64(smtLevel*2) {
+					e2eskipper.Skipf("Skipping CPU Manager %s tests since the CPU capacity < 4", fullCPUsOnlyOpt)
+				}
+
+			})
+
+			ginkgo.AfterEach(func(ctx context.Context) {
+				deletePodsAsync(ctx, f, podMap)
+			})
+
+			ginkgo.It("should reject workload asking non-SMT-multiple of cpus", func(ctx context.Context) {
+				reservedCPUs := cpuset.New(0)
+
+				// TODO: we assume the first available CPUID is 0, which is pretty fair, but we should probably
+				// check what we do have in the node.
+				updateKubeletConfigIfNeeded(ctx, f, configureCPUManagerInKubelet(oldCfg, &cpuManagerKubeletArguments{
+					policyName:              string(cpumanager.PolicyStatic),
+					reservedSystemCPUs:      reservedCPUs,
+					enableCPUManagerOptions: true,
+					options: map[string]string{
+						cpumanager.FullPCPUsOnlyOption: "true",
+					},
+				}))
+
+				ginkgo.By("creating the testing pod")
+				// negative test: try to run a container whose requests aren't a multiple of SMT level, expect a rejection
+				ctnAttrs := []ctnAttribute{
+					{
+						ctnName:    "gu-container-neg",
+						cpuRequest: "1000m",
+						cpuLimit:   "1000m",
+					},
+				}
+				pod := makeCPUManagerPod("gu-pod", ctnAttrs)
+				// CreateSync would wait for pod to become Ready - which will never happen if production code works as intended!
+				pod = e2epod.NewPodClient(f).Create(ctx, pod)
+				podMap[pod.Name] = pod
+
+				ginkgo.By("ensuring the testing pod is in failed state")
+				err := e2epod.WaitForPodCondition(ctx, f.ClientSet, f.Namespace.Name, pod.Name, "Failed", 30*time.Second, func(pod *v1.Pod) (bool, error) {
+					if pod.Status.Phase != v1.PodPending {
+						return true, nil
+					}
+					return false, nil
+				})
+				framework.ExpectNoError(err)
+
+				ginkgo.By("ensuring the testing pod is failed for the expected reason")
+				pod, err = e2epod.NewPodClient(f).Get(ctx, pod.Name, metav1.GetOptions{})
+				framework.ExpectNoError(err)
+
+				if pod.Status.Phase != v1.PodFailed {
+					framework.Failf("pod %s not failed: %v", pod.Name, pod.Status)
+				}
+				if !isSMTAlignmentError(pod) {
+					framework.Failf("pod %s failed for wrong reason: %q", pod.Name, pod.Status.Reason)
+				}
+			})
+
+			ginkgo.It("should admit workload asking SMT-multiple of cpus", func(ctx context.Context) {
+				// positive test: try to run a container whose requests are a multiple of SMT level, check allocated cores
+				// 1. are core siblings
+				// 2. take a full core
+				// WARNING: this assumes 2-way SMT systems - we don't know how to access other SMT levels.
```

Comment:
I wonder if it would be better for us to determine CPUs per core on the system and then use that when we specify the number of CPUs in the pod CPU request? It should be straightforward and would allow us to get rid of the assumption that the system has 2 CPUs on a core (i.e. 2 way SMT).